src/tools/rbd_mirror/instance_watcher/Types.h

```diff
@@ -39,8 +39,110 @@ WRITE_CLASS_ENCODER(PeerImageId);
 
 typedef std::set<PeerImageId> PeerImageIds;
 
+enum NotifyOp {
+  NOTIFY_OP_IMAGE_ACQUIRE  = 0,
+  NOTIFY_OP_IMAGE_RELEASE  = 1,
+};
+
+struct ImageAcquirePayload {
+  static const NotifyOp NOTIFY_OP = NOTIFY_OP_IMAGE_ACQUIRE;
+
+  uint64_t request_id;
+  std::string global_image_id;
+  PeerImageIds peers;
```

Comment:
I wonder if we can eliminate this from the acquire/release notifications. Each rbd-mirror daemon instance (currently) tracks its peers (albeit only a single peer right now). The leader instance doesn't know which peer (if any) are primary, so regardless it's up to the assigned instance's ImageReplayer to figure that out. Therefore, on acquire the peer instance should have enough information to start. On the release side, just because an image disappears doesn't necessarily imply the primary was deleted. Again, right now we have ```ImageDeleter``` checking to see if we should really remove the image. I am thinking the ```ImageDeleter``` hooks that currently exist within ```InstanceReplayer``` should be moved to ```ImageReplayer```. I'll try to work on this idea tomorrow in the current codebase.